### PR TITLE
Adjust embark-consult to handle suffix tofus

### PR DESCRIPTION
**Pull request by minad**
See https://github.com/minad/consult/commit/4d0a8b9b5a7f76e7055726fb79ed73d01a49aa01. Please give it a check!

### CODE REVIEW

**Comment by oantolin** [comment]
Looks good. I'll give it a test when I'm at the computer and then merge.

File: embark-consult.el [hunk 1]
```diff
@@ -137,7 +137,9 @@ associated to an active minibuffer for a Consult command."
 (defun embark-consult--strip-suffix (string property)
   "Remove the final characters of STRING with the given PROPERTY."
   (let ((len (length string)))
-    (substring string 0 (or (previous-single-property-change len property string) len))))
+    (if (and (> len 0) (get-text-property (- len 1) property string))
+        (substring string 0 (or (previous-single-property-change len property string) len))
+      string)))
```
Comment (oantolin) [inline embark-consult.el:142]: Saves an unnecessary copy, I like it. This is the sort of optimization that in Vertico you'd have to weight against bumping up the line count by two... 😛

Comment (minad) [reply embark-consult.el:142]: I object to that! Vertico is highly optimized, this is what recently bumped the line count to 498 lines (https://github.com/minad/vertico/commit/a962a8b2b682bb99304aa022bb20d66a838fd1ea). @manuel-uberti observes the line count closely. I have to act wisely with the remaining two lines if I want to keep the users.

Comment (oantolin) [reply embark-consult.el:142]: Is deferring the highlighting really noticeable? When Clemens asked for the skip-highlighting variable in orderless I remember testing with Selectrum and not noticing any difference.

**Comment by minad** [comment]
You can merge this as is.

But there is one small defect - `consult-global-mark` also uses `consult-location`. Before the suffix change the `consult-location` property was used to mark the part which should be stripped, in the case of `consult-global-mark` the `file:line:` location prefix. This will now break the embark occur export for `consult-global-mark`, which nobody uses.

Two alternatives:

1. Introduce a `consult-global-mark` category which will just be treated the same as the `consult-isearch` category, no occur export, only supports stripping the tofu suffix for `embark-save`.
2. Introduce another property `consult-location-strip` property which should be stripped by the `embark-consult--location-transform`.

I am inclined to go with alternative 1, what do you think?

**Comment by oantolin** [comment]
A third option is to go back to using `consult-location` to mark the parts that need to be stripped, but have the stripping function check both the beginning and the end of the string for it. But I agree that export for consult-global-mark is unlikely to be used.

**Comment by minad** [comment]
We wrote at the same time. See my edit above. I think option 2 is more general, so probably better in the long term. The problem with the option you propose is that I am always marking the beginning of the string with `consult-location`, even if the part should not be stripped.

**Comment by oantolin** [comment]
Export to occur mode with multiple buffers already worked, you can test it with `consult-global-mark`!

**Comment by minad** [comment]
Yes, I assumed that. Never tried and never used it, for obvious reasons :laughing: I will push a change for option 2 then!

**Comment by oantolin** [comment]
I say we go with your option 2, for the sake of a future `consult-line-multi-buffer`, and let's keep the export working for `consult-global-mark`, why not?

**Comment by minad** [comment]
See e21856cdfc88566cf873cf68c513d65b94bda4ca for the general version.

**Comment by oantolin** [comment]
That does look nicer, don't you think? The only downside is that in consult you have to mark something as both invisible and `consult-strip`, but that doesn't sound too bad to me.

**Comment by minad** [comment]
> That does look nicer, don't you think? The only downside is that in consult you have to mark something as both invisible and consult-strip, but that doesn't sound too bad to me.

Yes, it is better. And we can also use this in the future if we need the additional flexibility.

**Comment by jaor** [comment]
On Sun, Apr 11 2021, Daniel Mendler wrote:

> @minad commented on this pull request.

> I have to act wisely with the remaining two lines if I want to keep
> the users.

you have. i don't say much, but i'm watching ;) (vertico is working like
a charm for me so far, and switching to the native compilation branch
really made it faster, for some reason).

**Comment by oantolin** [comment]
I believe this is ready. Anything else you wanted to do before I merge, @minad?

**Comment by minad** [comment]
@oantolin It is ready! @jaor Thanks :)

**Comment by oantolin** [comment]
Merged. Thanks a lot for doing this in the first place, and for making the changes I suggested here and in Consult.

**Comment by manuel-uberti** [comment]
@minad `vertico` is awesome and I'm willing to bear the tragic loss of the 500 lines limit if it will eventually happen. ;)

**Comment by minad** [comment]
@manuel-uberti Thank you :D 

But staying minimal with Vertico is really good since it allows me to explore the limits of what the builtin completion infrastructure provides. There is a lot in there, and Vertico just uses it without inventing anything of its own. It performs a little bit of its own state management in contrast to Icomplete, but this is architecturally cleaner, simpler and more robust.

**Pull request merged**